### ISSUE TITLE
simplify remote plugins

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim#27949]
### Problem

The "remote plugin" model currently is too complex, for not much benefit.

- lifecycle is complicated.
- too many moving parts, too many concepts.
    - why is a "plugin host" needed?
    - why can't I just call `vim.rplugin('node', 'path/to/index.js')`, then call functions defined in the remote process?
    - why do I need "decorators" (`@pynvim.plugin`) ? why can't I just use Lua to define commands/events in `plugin/foo.lua`?
        - `@pynvim.command('MyCommand', …)` vs Lua `vim.rplugin(…):def_cmd('MyCommand', …)`

https://github.com/neovim/neovim/issues/5532#issuecomment-336474810

### Expected behavior

- from `doc/remote_plugin.txt`:
  > Plugin hosts ... take care of most boilerplate involved in defining
  > commands, autocmds, and functions implemented over |RPC| connections.
  > Hosts are loaded only when one of their registered plugins require it.
- address the above use-cases as follows:
  1. remote-plugin "main" is always a _Lua_ plugin (plugin/foo.lua).
      - Eliminates all the unnecessary redundant impls that try to find foo.js/foo.py/… plugin mains.
  2. make it easy from _Lua_ to start _any_ API client and treat it as a "remote host".
      - One process per plugin.
      - Eliminates all the redundant, per-platform (js/py/…) impls that implement a "remote host".
        Get rid of the concept of a "plugin host". We only need API clients, not any extra "plugin host" concept.
        A "plugin host" is just an API client started by Nvim, which will provide a Lua interface to:
          1. discover the "main" of a remote plugin and start its relevant API client.
          2. ergonomically define commands/autocmds/functions that call remote functions.
      - Example (compare to `:help remote-plugin-example`):
        - Python code ("remote"):
          ```
          import pynvim

          class Limit(object):
              def __init__(self, vim):
                  self.vim = vim
                  self.calls = 0
              ...

          plugin = Limit(object):
          ```
        - Lua code ("local"):
          ```
          local rplugin = vim.rplugin('python', 'path/to/init.py')

          vim.api.nvim_create_user_command('Cmd',
            function(args)
              vim.rpcrequest(rplugin.chan_id, 'command_handler', args.args, args.range)
            end,
            { nargs = '*' })

          vim.api.nvim_create_autocmd({'BufEnter'}, {
            pattern = {"*.c", "*.h"},
            callback = function(ev)
              vim.rpcrequest(rplugin.chan_id, 'autocmd_handler', ev.file)
            end
          })

          function Func(...)
            vim.rpcrequest(rplugin.chan_id, 'function_handler', {...})
          end
          ```
  3. With the above design...
      - ELIMINATES:
          - a fuckton of redundant documentation explaining crap like decorators, the "sync" flag, Remote plugin manifest, `remote#host#RegisterPlugin`, "bootstrapping" details.
          - a fuckton of code, e.g.: https://github.com/neovim/node-client/tree/master/packages/neovim/src/plugin/
          - `NvimPlugin`
          - `rplugin/node/` runtimepath directory
          - "registration"
          - `:UpdateRemotePlugins`
          - rplugin "manifest"
          - `remote#host#PluginsForHost`
          - `NvimPlugin.registerFunction`, `NvimPlugin.registerAutocmd`, `NvimPlugin.registerCommand`
      - STILL NEEDED:
          - instead of `NvimPlugin`, the remote plugin uses the exact same `NvimClient` type that is returned by `attach()`.
            A remote plugin is just a library that operates on the same old `NvimClient` that any other API client operates on.
          - `provider#Poll()`
          - `detect()`
          - `require()`
      - GAINS:
          - The "bootstrapping" is now extremely obvioius and the plugin implementor fully controls when to call `vim.rplugin()` from Lua.
          - `vim.rplugin()` loads the plugin "main" and returns a channel:
            1. find the platform interpreter (`node`, `python`, …)
            2. start the interpreter with a stdio RPC channel, targeting the plugin "main" file.
                - The plugin "main" file is expected to `attach()` to stdio, and use the resulting `NvimClient` to serve requests.
            3. calls `provider#Poll()` until success.
            4. returns a channel id.
          - NO sugar for creating commands/functions/autocmds that connect to the remote plugin.
            If creating commands/functions/autocmds is cumbersome we should fix that IN GENERAL, not only for remote plugins.
            ```
            vim.api.nvim_create_autocmd({'BufEnter'}, {
              pattern = {"*.c", "*.h"},
              callback = function(ev)
                vim.rpcrequest(rplugin.chan_id, 'autocmd_handler', ev.file)
              end
            })
            ```
  4. FAQ
      - How does a plugin avoid loading the interpreter (slow) on startup?
          - We could add an optional `ns` (namespace) parameter to `vim.rplugin()`, then it could be called on-demand and re-uses the existing channel if found.

**Comment by tarruda** [comment]
Hi @justinmk. Following up from your [comment](https://github.com/neovim/neovim/issues/4439#issuecomment-2009774868) in a issue I was subscribed to.


I will try to answer some of your questions, but please take it with a grain of salt as it has been 8 years since I worked on Neovim:


> Lifecycle is complicated, unnecessarily hard to test, troubleshoot, and explain to plugin authors.

Can you elaborate on this one? I don't remember and couldn't find anything in the docs.


> Every "remote plugin host" must implement a 1000s of lines of code to solve the same problem in every API client. This is a pointless waste of time. 
> why is a "plugin host" needed?


The idea of a plugin host is to implement the necessary boilerplate code to communicate with Neovim. There should not be any need to write more than one plugin host per language (at least that was my goal at the time).


> why can't I just call vim.rplugin('node', 'path/to/index.js'), then call functions defined in the remote process?

How would `path/to/index.js` expose functions? Neovim uses msgpack-rpc to do it, so at very least there should be some infrastructure that allows it to expose functions via RPC, which should be done by the plugin host.


> why do I need "decorators" (@pynvim.plugin) ?

IIRC, this decorator would be part of the python plugin host and an easy way to export functions to Neovim via RPC.


> why can't I just use Lua to define commands/events in plugin/foo.lua? 

Not sure if I understood this question, but if you are writing Lua plugins, then it is not a remote plugin.


In hindsight, the idea of remote plugins might not have been that great. While in theory the ability to write plugins in any programming language is cool, in practice what happened is that the community ended up embracing Lua as the standard language for plugins (At least I haven't seen any major plugins written in anything other than Lua and vimscript).

**Comment by justinmk** [comment]
> The idea of a plugin host is to implement the necessary boilerplate code to communicate with Neovim.

I am distinguishing because "API client" and "plugin host". 

- The "API client" part of node-client is here:
    - https://github.com/neovim/node-client/tree/master/packages/neovim/src/attach
    - https://github.com/neovim/node-client/tree/master/packages/neovim/src/api
- The "plugin host" part of node-client is here: 
    - https://github.com/neovim/node-client/tree/master/packages/neovim/src/host
    - https://github.com/neovim/node-client/tree/master/packages/neovim/src/plugin

The "plugin host" is 99% just nice-to-have "sugar" that makes it possible to define Nvim commands/autocmds from javascript. None of that is really necessary, it's easy to do that from Lua (or we should improve that).

> There should not be any need to write more than one plugin host per language

But implementing the "remote host" nice-to-have stuff in every language is a chore, and a source of extra bugs, and doubles the amount of code + docs needed in each API client.

> How would path/to/index.js expose functions?

The remote host will need to handle `rpcrequest()` calls from Nvim and "dispatch" them. That will require some sort of initialization/registration in the remote module. But that still avoids needing to register things in a manifest from the Nvim (client) side.

> IIRC, this decorator would be part of the python plugin host and an easy way to export functions to Neovim via RPC.

Yes. Not worth the complexity and bugs.

> Not sure if I understood this question, but if you are writing Lua plugins, then it is not a remote plugin.

I'm proposing to move the _bindings_ logic to Lua:

- start and manage an API client(s), from Lua
- call remote functions, and setup remote handlers, from Lua

but the _business logic_ stays in the remote module.

Currently, trying to set up _bindings_ from the remote side is 99% of the complexity, but 1% of the value. The business logic (not bindings) is where remote plugins can be useful. Furthermore, with this simplified design, we get it nearly _for free_.

> in practice what happened is that the community ended up embracing Lua 

That's parially because creating a remote plugin is too complicated. This proposal fixes that. There are cases where I want to integrate with code that doesn't have a CLI but can be loaded as a node/python/etc module.

**Comment by justinmk** [comment]
Skeleton / proof of concept: https://github.com/neovim/node-client/pull/344

**Comment by dundargoc** [comment]
i ain’t reading all that. i'm happy for you tho. or sorry that happened

**Comment by brianhuster** [comment]
So `vim.rplugin` hasn't been implemented yet, right? I can't find it in the doc

**Comment by justinmk** [comment]
> So `vim.rplugin` hasn't been implemented yet, right?

Correct, and it may not be needed for the first phase (if ever), it's just pseudocode.

**Comment by dlants** [comment]
Hey, I ran into this when starting my node-based plugin.

I want to chime in and say that there's another benefit of implementing things this way, which is that it opens up the possibility of using alternate js runtimes - like deno and bun. Bun in particular claims much faster startup times and less memory overhead.

I did try to set things up in a way similar to this by setting up an init.lua entrypoint on the lua side with:
```
vim.fn.jobstart({
    'bun', 'your-script.ts',
    socket  -- Pass as argument to your script
  }, {
    env = {
      NVIM_LISTEN_ADDRESS = socket
    }
  })
```

and then using node client's "attach" mechanism on the node side.

```
  const nvim = attach({ socket: process.env.NVIM_LISTEN_ADDRESS });
```

This got me the ability to manipulate nvim from bun. However, I got stuck on how to trigger bun code from nvim. So how to define a command for example.

I think there is an rpc channel from nvim to node, so it should be possible to send things along it from the lua side and receive it on the node end.

Overall this felt like a yak-shaving exercise for the actual plugin development so I didn't dig into it beyond this. But it would be nice to know how to set this up in a "future proof" way. Thanks, I appreciate all your work on nvim!

**Comment by justinmk** [comment]
> However, I got stuck on how to trigger bun code from nvim. So how to define a command for example.
> 
> I think there is an rpc channel from nvim to node, so it should be possible to send things along it from the lua side and receive it on the node end.

@dlants yup, the only thing missing is https://github.com/neovim/node-client/pull/344 which would allow your js code to define a handler.

**Comment by brianhuster** [comment]
Btw, we can use `vim.fn.py3eval` to call Python functions from Lua code. Is there similar thing in other clients?
